sw/device/lib/dif/dif_i2c.c

```diff
@@ -0,0 +1,624 @@
+                                        .mask = I2C_TIMING4_T_BUF_MASK,
+                                        .index = I2C_TIMING4_T_BUF_OFFSET,
+                                        .value = timing.stop_signal_hold_time,
+                                    });
+  mmio_region_write32(i2c->base_addr, I2C_TIMING4_REG_OFFSET, timing4);
+}
+
+dif_i2c_result_t dif_i2c_init(mmio_region_t base_addr,
+                              dif_i2c_timing_params_t timing, dif_i2c_t *i2c) {
+  if (i2c == NULL) {
+    return kDifI2cResultBadArg;
+  }
+
+  i2c->base_addr = base_addr;
+
+  // Turn the device off, before resetting it.
+  dif_i2c_result_t error;
+  error = dif_i2c_host_set_enabled(i2c, kDifI2cDisabled);
+  if (error != kDifI2cResultOk) {
+    return error;
+  }
+
+  // Clear + disable all interrupts. We do this directly, rather than using the
+  // actual DIF functions, to perform both in single store operations.
+  mmio_region_write32(i2c->base_addr, I2C_INTR_STATE_REG_OFFSET, UINT32_MAX);
+  mmio_region_write32(i2c->base_addr, I2C_INTR_ENABLE_REG_OFFSET, 0);
+
+  // Reset all FIFO state.
+  error = dif_i2c_reset_rx_fifo(i2c);
+  if (error != kDifI2cResultOk) {
+    return error;
+  }
+  error = dif_i2c_reset_fmt_fifo(i2c);
+  if (error != kDifI2cResultOk) {
+    return error;
+  }
+  error = dif_i2c_set_watermarks(i2c, kDifI2cLevel1Byte, kDifI2cLevel1Byte);
+  if (error != kDifI2cResultOk) {
+    return error;
+  }
+
+  write_timing_params(i2c, timing);
+
+  return kDifI2cResultOk;
+}
+
+dif_i2c_result_t dif_i2c_reset_rx_fifo(const dif_i2c_t *i2c) {
+  if (i2c == NULL) {
+    return kDifI2cResultBadArg;
+  }
+
+  mmio_region_nonatomic_set_bit32(i2c->base_addr, I2C_FIFO_CTRL_REG_OFFSET,
+                                  I2C_FIFO_CTRL_RXRST);
+
+  return kDifI2cResultOk;
+}
+
+dif_i2c_result_t dif_i2c_reset_fmt_fifo(const dif_i2c_t *i2c) {
+  if (i2c == NULL) {
+    return kDifI2cResultBadArg;
+  }
+
+  mmio_region_nonatomic_set_bit32(i2c->base_addr, I2C_FIFO_CTRL_REG_OFFSET,
+                                  I2C_FIFO_CTRL_FMTRST);
+
+  return kDifI2cResultOk;
+}
+
+dif_i2c_result_t dif_i2c_set_watermarks(const dif_i2c_t *i2c,
+                                        dif_i2c_level_t rx_level,
+                                        dif_i2c_level_t fmt_level) {
+  if (i2c == NULL) {
+    return kDifI2cResultBadArg;
+  }
+
+  ptrdiff_t rx_level_value;
+  switch (rx_level) {
+    case kDifI2cLevel1Byte:
+      rx_level_value = I2C_FIFO_CTRL_RXILVL_RXLVL1;
+      break;
+    case kDifI2cLevel4Byte:
+      rx_level_value = I2C_FIFO_CTRL_RXILVL_RXLVL4;
+      break;
+    case kDifI2cLevel8Byte:
+      rx_level_value = I2C_FIFO_CTRL_RXILVL_RXLVL8;
+      break;
+    case kDifI2cLevel16Byte:
+      rx_level_value = I2C_FIFO_CTRL_RXILVL_RXLVL16;
+      break;
+    case kDifI2cLevel30Byte:
+      rx_level_value = I2C_FIFO_CTRL_RXILVL_RXLVL30;
+      break;
+    default:
+      return kDifI2cResultBadArg;
```

Comment:
Is this supposed to be `kDifI2cError`?

Reply:
No? Error is only returned after you've touched MMIO registers.